libcache/cache.c

```diff
@@ -0,0 +1,696 @@
+#include <errno.h>
+#include <string.h>
+
+
+#define LIBCACHE_NUM_WAYS   4
+#define LIBCACHE_ADDR_WIDTH 64
+
+
+#define LIBCACHE_TIMESTAMPS_UPDATE    1
+#define LIBCACHE_TIMESTAMPS_NO_UPDATE 0
+
+
+#define IS_VALID(f) (((f) & (1 << 0)) != 0 ? 1 : 0)
+#define IS_DIRTY(f) (((f) & (1 << 1)) != 0 ? 1 : 0)
+
+#define SET_VALID(f) \
+	do { \
+		(f) |= (1 << 0); \
+	} while (0)
+#define CLEAR_VALID(f) \
+	do { \
+		(f) &= ~(1 << 0); \
+	} while (0)
+
+#define SET_DIRTY(f) \
+	do { \
+		(f) |= (1 << 1); \
+	} while (0)
+#define CLEAR_DIRTY(f) \
+	do { \
+		(f) &= ~(1 << 1); \
+	} while (0)
+
+
+#define LOG2(x) ((uint8_t)(8 * sizeof(unsigned long) - __builtin_clzl((x)) - 1))
+
+
+typedef struct cacheline_s cacheline_t;
+
+
+struct cacheline_s {
+	uint64_t tag;
+	cacheline_t *prev, *next; /* Circular doubly linked list */
+	void *data;
+	unsigned char flags;
+};
+
+
+typedef struct {
+	cacheline_t *timestamps;
+	cacheline_t *tags[LIBCACHE_NUM_WAYS];
+	cacheline_t lines[LIBCACHE_NUM_WAYS];
+	size_t count;
+} cacheset_t;
+
+
+struct cachectx_s {
+	cacheset_t *sets;
+
+	size_t srcMemSize;
+	size_t lineSize;
+	size_t numLines;
+	size_t numSets;
+
+	uint64_t tagMask;
+	uint64_t setMask;
+	uint64_t offMask;
+
+	uint8_t offBitsNum;
+	uint8_t setBitsNum;
+	uint8_t tagBitsNum;
+
+	cache_readCb_t readCb;
+	cache_writeCb_t writeCb;
+
+	handle_t lock;
+};
+
+
+static uint64_t cache_generateMask(int numBits)
+{
+	return ((uint64_t)1 << numBits) - 1;
+}
+
+
+cachectx_t *cache_init(size_t srcMemSize, size_t size, size_t lineSize, cache_writeCb_t writeCb, cache_readCb_t readCb)
+{
+	int err;
+	cachectx_t *cache = NULL;
+
+	if (srcMemSize == 0 || size == 0 || lineSize == 0) {
+		cache = NULL;
+	}
+	else if (size % lineSize != 0 || (size / lineSize) % LIBCACHE_NUM_WAYS != 0) {
+		cache = NULL;
+	}
+	else {
+		cache = malloc(sizeof(cachectx_t));
+
+		if (cache != NULL) {
+			cache->srcMemSize = srcMemSize;
+			cache->lineSize = lineSize;
+			cache->numLines = size / cache->lineSize;
+
+			cache->numSets = cache->numLines / LIBCACHE_NUM_WAYS;
+
+			cache->sets = calloc(cache->numSets, sizeof(cacheset_t));
+
+			if (cache->sets != NULL) {
+				cache->offBitsNum = LOG2(cache->lineSize);
+				cache->setBitsNum = LOG2(cache->numSets);
+				cache->tagBitsNum = LIBCACHE_ADDR_WIDTH - cache->setBitsNum - cache->offBitsNum;
+
+				cache->tagMask = cache_generateMask(cache->tagBitsNum);
+				cache->setMask = cache_generateMask(cache->setBitsNum);
+				cache->offMask = cache_generateMask(cache->offBitsNum);
+
+				cache->readCb = readCb;
+				cache->writeCb = writeCb;
+
+				err = mutexCreate(&cache->lock);
+				if (err < 0) {
+					free(cache->sets);
+					free(cache);
+					cache = NULL;
+				}
+			}
+			else {
+				free(cache);
+				cache = NULL;
+			}
+		}
+	}
+
+	return cache;
+}
+
+
+static uint64_t cache_computeAddr(const cachectx_t *cache, uint64_t tag, uint64_t setIndex)
+{
+	uint64_t temp = (tag << cache->setBitsNum) | setIndex;
+	return temp <<= cache->offBitsNum;
```

Comment:
Undesired and unnecessary side effect
```suggestion
	return temp << cache->offBitsNum;
```